keps/sig-scheduling/2249-pod-affinity-namespace-selector/README.md

```diff
@@ -0,0 +1,734 @@
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+I am running a SaaS service, the workload for each customer is placed in a separate
+namespace. The workloads requires 1:1 pod to node placement. I want to use pod 
+anti-affinity across all customers namespaces to achieve that.
+
+<!--
+
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Performance
+Using namespace selector will make it easier for users to specify affinity 
+constraints across a large number of namespaces. The initial implementation 
+of pod affinity/anti-affinity suffered from performance challenges, however 
+over the releases 1.17 - 1.20 we significantly improved that. 
+
+We currently have integration and clusterloader benchmarks that evaluate the extreme
+cases of all pods having affinity/anti-affinity constraints to/against each other.
+Those benchmarks show that the scheduler is able to achieve maximum throughput 
+(i.e., the api-server qps limit).
+
+#### Security/Abuse
+See previous discussion [here](https://github.com/kubernetes/kubernetes/issues/68827#issuecomment-468470598)
+NamespaceSelector will allow users to select all namespaces. This may cause a security 
+concern: a pod with anti-affinity constraint can block pods from all other namespaces 
+from getting scheduled in a failure domain.
+
+We will address this concern by introducing a new quota scope named `CrossNamespaceAffinity`
+that admins can use to limit which namespaces are allowed to have pods with affinity terms
+that set the existing `namespaces` field or the proposed one `namespaceSelector`. 
+
+Using this new scope, admins can prevent certain namespaces from having pods that use
+cross-namespace pod affinity by creating a resource quota object in that namespace with
+`CrossNamespaceAffinity` scope and hard limit of 0:
+
+```go
+apiVersion: v1
+kind: ResourceQuota
+metadata:
+    name: pods-medium
+spec:
+    hard:
+      pods: "0"
+    scopeSelector:
+      matchExpressions:
+      - scopeName: CrossNamespaceAffinity
+```
+
+
+If admins want to disallow using `namespaces` and `namespaceSelector` by default, and 
```

Comment:
admins or providers?
That is, do admins need access to the control plane?

Reply:
admins. 
They don't need access to the control plane.

Reply:
changed to use a more neutral term: operators.